### ISSUE TITLE
Is there any option to translate slugs as well?

### ISSUE THREAD

**Issue by mgarciadelojo** [mcamara/laravel-localization#303]
Hi,

I am testing this package and I find it awesome. Before knowing its existence, I tried to do it myself and my solution was very close to it, so, ... to keep my code clean and don't reinvent the wheel I 'll use it instead of my own code.

The thing is, in my database I have two slug fields for spanish and english named as follows:

- slug_es
- slug_en

I followed the docs, and obviously the package takes always the slug in the current language because it takes the route attributes.

How can I change the route attributes in my controllers?

Any idea would be very helpful.

Many thanks.

**Comment by mgarciadelojo** [comment]
Well, what I have done is, passing the route attributes to my layout template where the language switcher is rendering.

For example, in one `show.blade.php`I have:

```
@extends('layouts.app', [
    'route_attributes' => [
        'route_name' => 'routes.showPost',
        'es' => ['slug' => $post->slug_es],
        'en' => ['slug' => $post->slug_en],
    ],
])
```

And in the `app.blade.php`...

```
@foreach(LaravelLocalization::getSupportedLocales() as $localeCode => $properties)
    {{--*/
        $localizedUrl = isset($route_attributes)
            ? LaravelLocalization::getURLFromRouteNameTranslated(
                $localeCode,
                $route_attributes['route_name'],
                $route_attributes[$localeCode])
            : LaravelLocalization::getLocalizedURL($localeCode)
    /*--}}
    <li>
        <a rel="alternate" hreflang="{{ $localeCode }}" href="{{ $localizedUrl }}">
            <span class="flag_{{ $localeCode }}"></span> {{ $properties['native'] }}
        </a>
    </li>
@endforeach
```

I don't like to be defining this in every single view, but, at least my controllers don't get dirty.

Hope this helps many of you.

**Comment by BojanKogoj** [comment]
I'm having the same problem, unfortunately your solution won't work for me. If I figure it out I'll let you know

**Comment by mgarciadelojo** [comment]
Could you provide more info or code example?

**Comment by rtucek** [comment]
@mgarciadelojo what do you exactly mean by:
> How can I change the route attributes in my controllers?

It sound's like you would like to alter the route on runtime (after the framework's bootstrapping process).

**Comment by mgarciadelojo** [comment]
Yes, that's it, but I found out another solution I shared above :)

**Comment by rtucek** [comment]
TL;DR
A controller is the answer to a (resolved) route (from a HTTP request) and not meant for defining routes.

And the full explanation...
It's technically possible to declare anywhere a route (just so by using the Route facade), however that's a very, very, very bad practice and it's not the way how Laravel should be used. There are only 2 places, where you should place your routes:

1. **If a route is part of your Laravel app**, than keep it in the Laravel's routes.php (using the common [Route facade](https://laravel.com/docs/master/routing)).
2. **If (and only if) you're route is a part of a service container**, than you can define the route within the boot method of the service provider. This is what [Laravel package developer](https://laravel.com/docs/5.2/packages#routing) should do.

You may ask yourself, why a route should not be defined like the way as described for package devs IN a controller?
That's simple: on every page request, any framework (Wordpress, Laravel, Drupal, Symphony, CodeIgnitor and whatsoever) is loading it's own core source code files and preparing a proper environment and API for developers to interact with. Such a process is called bootstrapping. At the very end of the bootstrapping process, Laravel is loading the routes file and all service provider, defined in your config/app.php. There is actually a lot going on before anything from your source code will be executed, but it's just happening in a blink of an eye.

So when the framework is ready and your routes are loaded in place, Laravel's router will dispatch the request, according to your previously defined routes, to a specific controller (more precise to a specific controller method).

So if you've read my words closely you may understand why it's useless to define a route in your Controller. A controller is never ever part of a bootstrapping process (no matter how many you have, just one of them will be executed by the router).
See [request lifecycle](https://laravel.com/docs/5.2/lifecycle).

I guess your controller is handling some logic and will decide if a user is permitted to access a route or not (right?). If that's the case, than you could use so many different approaches. One quick working solution is to attach a middleware to the route and hardcode your logic there if a client is permitted to proceed further OR should be redirected somewhere else... yeah, truly a quick solution but ugly and not sustainable.
The best way is to look at those pages in the documentation [authentication](https://laravel.com/docs/5.2/authentication) and [authorization](https://laravel.com/docs/master/authorization).

Or are you trying to achieve something else?

**Comment by mgarciadelojo** [comment]
Hi @rtucek, and thanks for the great explanation,

Unfortunately, I think you didn't understand my question well. If you read it carefully, I was trying to ask how to transalate slugs for items in my database. That's what this package calls **route attributes** if you look at the code of the class `LaravelLocalization.php`.

What I did to solve this situation, is passing to my layout both slugs, so, if a variable `route_attributes` exists in the layout, then, the language switcher will use it and call the function `getURLFromRouteNameTranslated` instead of `getLocalizedURL`.

I do understand what you were trying to explain, and really appreciate it. Maybe will help other devs to understand the laravel's bootstrapping process.

Cheers,
